### PR TITLE
Configure Address using NLog Layout. Support for ThreadId

### PR DESCRIPTION
**Pull request by snakefoot**
Apply ContextProperties for capturing logging context, so ThreadId is correctly captured.

NLog CallSite-capture can be enabled with the option `IncludeCallSite=true` (Introduces performance hit).

```xml
<nlog>
 <extensions>
   <add assembly="NLog.Web.AspNetCore"/>
   <add assembly="Analogy.LogViewer.NLog.Targets"/>
 </extensions>

 <targets>
  <!-- write logs to Analogy Log Server  -->
  <target xsi:type="NlogAnalogyGRPCTarget" name="Analogy" address="http://localhost:6000" layout="${message}${exception:format=tostring}" includeCallSite="true">
      <contextproperty name="UserName" layout="${aspnet-user-identity}" />
      <contextproperty name="MachineName" layout="${machinename}" />
      <contextproperty name="ProcessName" layout="${processmame}" />
      <contextproperty name="ThreadId" layout="${threadid}" />
      <contextproperty name="Source" layout="${logger}" />
      <contextproperty name="Category" layout="${exception:format=shorttype}" />
  </target>
 </targets>

 <rules>
   <logger name="*" minlevel="Trace" writeTo="Analogy" />
 </rules>
</nlog>
```

Instead of having these hardcoded properties in the producer-log-interface, then it would be easier if the interface was just two parameters `IReadOnlyDictionary<string, string> defaultColumns` and `IReadOnlyDictionary<string, string> additionalDetails`. Then the logger can decide if wanting to fill out all recognized defaultColumns (ProcessId, ThreadId, etc.) or add new colums.

### CODE REVIEW

File: Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs [hunk 1]
```diff
@@ -12,56 +14,126 @@ namespace Analogy.LogViewer.NLog.Targets
...
+            for (int i = 0; i < ContextProperties.Count; ++i)
+            {
+                var contextProperty = ContextProperties[i];
+                if (Constants.MachineName.Equals(contextProperty.Name))
+                    machineName = RenderLogEvent(contextProperty.Layout, logEvent);
+                else if (Constants.ProcessName.Equals(contextProperty.Name))
+                    processName = RenderLogEvent(contextProperty.Layout, logEvent);
+                else if (Constants.UserName.Equals(contextProperty.Name))
+                    userName = RenderLogEvent(contextProperty.Layout, logEvent);
+                else if (Constants.Category.Equals(contextProperty.Name))
```
Comment (LiorBanai) [inline Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs:79]: Codacy found an issue: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/gh/Analogy-LogViewer/Analogy.LogViewer.NLog.Targets/pullRequest?prid=6366134)

**Comment by LiorBanai** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 16
           

Complexity increasing per file
==============================
- Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Analogy-LogViewer/Analogy.LogViewer.NLog.Targets/pullRequest?prid=6366134&bid=20659658)

File: Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs [hunk 2]
```diff
@@ -12,56 +14,126 @@ namespace Analogy.LogViewer.NLog.Targets
...
+                    categoryName = RenderLogEvent(contextProperty.Layout, logEvent);
+                else if (Constants.Source.Equals(contextProperty.Name))
+                    sourceName = RenderLogEvent(contextProperty.Layout, logEvent);
+                else if (Constants.ThreadId.Equals(contextProperty.Name))
+                {
+                    var threadIdProperty = RenderLogEvent(contextProperty.Layout, logEvent);
+                    if (!string.IsNullOrEmpty(threadIdProperty))
```
Comment (LiorBanai) [inline Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs:86]: Codacy found an issue: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/gh/Analogy-LogViewer/Analogy.LogViewer.NLog.Targets/pullRequest?prid=6366134)

File: Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs [hunk 3]
```diff
@@ -12,56 +14,126 @@ namespace Analogy.LogViewer.NLog.Targets
...
+                }
+                else if (Constants.ProcessId.Equals(contextProperty.Name))
+                {
+                    var processIdProperty = RenderLogEvent(contextProperty.Layout, logEvent);
+                    if (!string.IsNullOrEmpty(processIdProperty))
+                        int.TryParse(processIdProperty, out processId);
+                }
+            }
+
+            // var additionalInformation = GetAllProperties();
```
Comment (LiorBanai) [inline Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs:97]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/gh/Analogy-LogViewer/Analogy.LogViewer.NLog.Targets/pullRequest?prid=6366134)

File: Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs [hunk 4]
```diff
@@ -12,56 +14,126 @@ namespace Analogy.LogViewer.NLog.Targets
...
         {
         }
+
         public NLogToAnalogyGRPCTarget(string address)
         {
-            producer = new LogServer.Clients.AnalogyMessageProducer(address, null);
+            Address = address;
+            Layout = "${message}";
+
+            // IncludeCallSite = true;
```
Comment (LiorBanai) [inline Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs:32]: Codacy found an issue: [Remove this commented out code.](https://app.codacy.com/gh/Analogy-LogViewer/Analogy.LogViewer.NLog.Targets/pullRequest?prid=6366134)

File: Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs [hunk 5]
```diff
@@ -12,56 +14,126 @@ namespace Analogy.LogViewer.NLog.Targets
...
+            string userName = null;
+            string categoryName = string.Empty;
+            for (int i = 0; i < ContextProperties.Count; ++i)
+            {
+                var contextProperty = ContextProperties[i];
+                if (Constants.MachineName.Equals(contextProperty.Name))
+                    machineName = RenderLogEvent(contextProperty.Layout, logEvent);
+                else if (Constants.ProcessName.Equals(contextProperty.Name))
+                    processName = RenderLogEvent(contextProperty.Layout, logEvent);
+                else if (Constants.UserName.Equals(contextProperty.Name))
```
Comment (LiorBanai) [inline Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs:77]: Codacy found an issue: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/gh/Analogy-LogViewer/Analogy.LogViewer.NLog.Targets/pullRequest?prid=6366134)

File: Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs [hunk 6]
```diff
@@ -12,56 +14,126 @@ namespace Analogy.LogViewer.NLog.Targets
...
+            int threadId = 0;
+            string sourceName = logEvent.LoggerName;
+            string machineName = null;
+            string processName = null;
+            string userName = null;
+            string categoryName = string.Empty;
+            for (int i = 0; i < ContextProperties.Count; ++i)
+            {
+                var contextProperty = ContextProperties[i];
+                if (Constants.MachineName.Equals(contextProperty.Name))
```
Comment (LiorBanai) [inline Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs:73]: Codacy found an issue: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/gh/Analogy-LogViewer/Analogy.LogViewer.NLog.Targets/pullRequest?prid=6366134)

File: Analogy.LogViewer.NLog.Targets/Constants.cs [hunk 7]
```diff
@@ -0,0 +1,13 @@
+﻿namespace Analogy.LogViewer.NLog.Targets
+{
+    public static class Constants
+    {
+        public const string ProcessId = "ProcessId";
+        public const string MachineName = "MachineName";
+        public const string ProcessName = "ProcessName";
```
Comment (LiorBanai) [inline Analogy.LogViewer.NLog.Targets/Constants.cs:7]: Codacy found an issue: [Change this constant to a 'static' read-only property.](https://app.codacy.com/gh/Analogy-LogViewer/Analogy.LogViewer.NLog.Targets/pullRequest?prid=6366134)

File: Analogy.LogViewer.NLog.Targets/Constants.cs [hunk 8]
```diff
@@ -0,0 +1,13 @@
...
+        public const string UserName = "UserName";
+        public const string Category = "Category";
```
Comment (LiorBanai) [inline Analogy.LogViewer.NLog.Targets/Constants.cs:9]: Codacy found an issue: [Change this constant to a 'static' read-only property.](https://app.codacy.com/gh/Analogy-LogViewer/Analogy.LogViewer.NLog.Targets/pullRequest?prid=6366134)

File: Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs [hunk 9]
```diff
@@ -12,56 +14,126 @@ namespace Analogy.LogViewer.NLog.Targets
...
+                    machineName = RenderLogEvent(contextProperty.Layout, logEvent);
+                else if (Constants.ProcessName.Equals(contextProperty.Name))
+                    processName = RenderLogEvent(contextProperty.Layout, logEvent);
+                else if (Constants.UserName.Equals(contextProperty.Name))
+                    userName = RenderLogEvent(contextProperty.Layout, logEvent);
+                else if (Constants.Category.Equals(contextProperty.Name))
+                    categoryName = RenderLogEvent(contextProperty.Layout, logEvent);
+                else if (Constants.Source.Equals(contextProperty.Name))
```
Comment (LiorBanai) [inline Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs:81]: Codacy found an issue: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/gh/Analogy-LogViewer/Analogy.LogViewer.NLog.Targets/pullRequest?prid=6366134)

File: Analogy.LogViewer.NLog.Targets/Constants.cs [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
+        public const string Source = "Source";
+        public const string ThreadId = "ThreadId";
```
Comment (LiorBanai) [inline Analogy.LogViewer.NLog.Targets/Constants.cs:11]: Codacy found an issue: [Change this constant to a 'static' read-only property.](https://app.codacy.com/gh/Analogy-LogViewer/Analogy.LogViewer.NLog.Targets/pullRequest?prid=6366134)

File: Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs [hunk 11]
```diff
@@ -12,56 +14,126 @@ namespace Analogy.LogViewer.NLog.Targets
...
+                else if (Constants.ThreadId.Equals(contextProperty.Name))
+                {
+                    var threadIdProperty = RenderLogEvent(contextProperty.Layout, logEvent);
+                    if (!string.IsNullOrEmpty(threadIdProperty))
+                        int.TryParse(threadIdProperty, out threadId);
+                }
+                else if (Constants.ProcessId.Equals(contextProperty.Name))
+                {
+                    var processIdProperty = RenderLogEvent(contextProperty.Layout, logEvent);
+                    if (!string.IsNullOrEmpty(processIdProperty))
```
Comment (LiorBanai) [inline Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs:92]: Codacy found an issue: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/gh/Analogy-LogViewer/Analogy.LogViewer.NLog.Targets/pullRequest?prid=6366134)

File: Analogy.LogViewer.NLog.Targets/Constants.cs [hunk 12]
```diff
@@ -0,0 +1,13 @@
+﻿namespace Analogy.LogViewer.NLog.Targets
+{
+    public static class Constants
+    {
+        public const string ProcessId = "ProcessId";
```
Comment (LiorBanai) [inline Analogy.LogViewer.NLog.Targets/Constants.cs:5]: Codacy found an issue: [Change this constant to a 'static' read-only property.](https://app.codacy.com/gh/Analogy-LogViewer/Analogy.LogViewer.NLog.Targets/pullRequest?prid=6366134)

File: Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs [hunk 13]
```diff
@@ -12,56 +14,126 @@ namespace Analogy.LogViewer.NLog.Targets
...
+            string machineName = null;
+            string processName = null;
+            string userName = null;
+            string categoryName = string.Empty;
+            for (int i = 0; i < ContextProperties.Count; ++i)
+            {
+                var contextProperty = ContextProperties[i];
+                if (Constants.MachineName.Equals(contextProperty.Name))
+                    machineName = RenderLogEvent(contextProperty.Layout, logEvent);
+                else if (Constants.ProcessName.Equals(contextProperty.Name))
```
Comment (LiorBanai) [inline Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs:75]: Codacy found an issue: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/gh/Analogy-LogViewer/Analogy.LogViewer.NLog.Targets/pullRequest?prid=6366134)

File: Analogy.LogViewer.NLog.Targets/Constants.cs [hunk 14]
```diff
@@ -0,0 +1,13 @@
...
+        public const string MachineName = "MachineName";
+        public const string ProcessName = "ProcessName";
+        public const string UserName = "UserName";
```
Comment (LiorBanai) [inline Analogy.LogViewer.NLog.Targets/Constants.cs:8]: Codacy found an issue: [Change this constant to a 'static' read-only property.](https://app.codacy.com/gh/Analogy-LogViewer/Analogy.LogViewer.NLog.Targets/pullRequest?prid=6366134)

File: Analogy.LogViewer.NLog.Targets/Constants.cs [hunk 15]
```diff
@@ -0,0 +1,13 @@
+﻿namespace Analogy.LogViewer.NLog.Targets
+{
+    public static class Constants
+    {
+        public const string ProcessId = "ProcessId";
+        public const string MachineName = "MachineName";
```
Comment (LiorBanai) [inline Analogy.LogViewer.NLog.Targets/Constants.cs:6]: Codacy found an issue: [Change this constant to a 'static' read-only property.](https://app.codacy.com/gh/Analogy-LogViewer/Analogy.LogViewer.NLog.Targets/pullRequest?prid=6366134)

File: Analogy.LogViewer.NLog.Targets/Constants.cs [hunk 16]
```diff
@@ -0,0 +1,13 @@
...
+        public const string ProcessName = "ProcessName";
+        public const string UserName = "UserName";
+        public const string Category = "Category";
+        public const string Source = "Source";
```
Comment (LiorBanai) [inline Analogy.LogViewer.NLog.Targets/Constants.cs:10]: Codacy found an issue: [Change this constant to a 'static' read-only property.](https://app.codacy.com/gh/Analogy-LogViewer/Analogy.LogViewer.NLog.Targets/pullRequest?prid=6366134)

**Comment by LiorBanai** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Issues
======
- Added 1
           

Complexity increasing per file
==============================
- Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Analogy-LogViewer/Analogy.LogViewer.NLog.Targets/pullRequest?prid=6366134&bid=20659658)

File: Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs [hunk 17]
```diff
@@ -12,56 +14,138 @@ namespace Analogy.LogViewer.NLog.Targets
...
+                    categoryName = RenderLogEvent(contextProperty.Layout, logEvent);
+                }
+                else if (Constants.Source.Equals(contextProperty.Name))
+                {
+                    sourceName = RenderLogEvent(contextProperty.Layout, logEvent);
+                }
+                else if (Constants.ThreadId.Equals(contextProperty.Name))
+                {
+                    var threadIdProperty = RenderLogEvent(contextProperty.Layout, logEvent);
+                    if (!string.IsNullOrEmpty(threadIdProperty))
```
Comment (LiorBanai) [inline Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs:96]: Codacy found an issue: [Add curly braces around the nested statement(s) in this 'if' block.](https://app.codacy.com/gh/Analogy-LogViewer/Analogy.LogViewer.NLog.Targets/pullRequest?prid=6366134)

**Comment by LiorBanai** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs  4
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Analogy-LogViewer/Analogy.LogViewer.NLog.Targets/pullRequest?prid=6366134&bid=20659658)

**Comment by LiorBanai** [comment]
Wow, Thanks for the improvements!

**Comment by LiorBanai** [comment]
<contextproperty name="ProcessName" layout="${processmame}" />

isn't this a typo?  processmame

**Comment by LiorBanai** [comment]
Codacy Here is an overview of what got changed by this pull request:


```diff

Complexity increasing per file
==============================
- Analogy.LogViewer.NLog.Targets/NLogToAnalogyGRPCTarget.cs  6
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/Analogy-LogViewer/Analogy.LogViewer.NLog.Targets/pullRequest?prid=6366134&bid=20659658)

**Comment by snakefoot** [comment]
Yes. It is a typo in the example config. Was just trying to show how to override the context values (If no context property then it will use default)


Sent from my Sony Xperia


---- Lior Banai wrote ----


isn't this a typo? processmame

**Comment by LiorBanai** [comment]
@snakefoot  Ok :).
Thanks again for the improvements!

**Comment by snakefoot** [comment]
> isn't this a typo? processmame

Fixed the typo

**Comment by LiorBanai** [comment]
Thanks.
I also considering adding: NLOG “real time” option https://github.com/nlog/nlog/wiki/NLogViewer-target to Analogy Log Viewer itself.

have you used it? is it usefull? the current Target is using gRPC which is netcore3.1 .. the  NLogViewer can be used in .net framework..

**Comment by LiorBanai** [comment]
maybe to my NLog repo : https://github.com/Analogy-LogViewer/Analogy.LogViewer.NLog

**Comment by snakefoot** [comment]
[NLogViewer-target](https://github.com/nlog/nlog/wiki/NLogViewer-target) uses the Log4jXml-format. So it would be a matter of creating an Analogy Log Viewer that supports Log4jXml-format using TCP / UDP. Could probably also support log4net Log4jXml-appender and ofcourse Java-Log4j-appenders. Maybe call it **Analogy.LogViewer.Log4jXml**

**Comment by snakefoot** [comment]
> the current Target is using gRPC which is netcore3.1 

Would be nice if it support NetStandard2.0 by depending on: https://www.nuget.org/packages/Grpc.Core.Api/

Then it would automatically support NetFramework.

**Comment by LiorBanai** [comment]
interesting :)
Thanks. I'll check those

**Pull request merged**